src/Microsoft.ML.SearchSpace/IParameter.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using Microsoft.ML.SearchSpace.Converter;
+
+namespace Microsoft.ML.SearchSpace
+{
+    public enum ParameterType
+    {
+        Integer = 0,
+        Float = 1,
+        Bool = 2,
+        String = 3,
+        Object = 4,
+        Array = 5,
+    }
+
+    [JsonConverter(typeof(ParameterConverter<IParameter>))]
+    public interface IParameter : IDictionary<string, IParameter>
```

Comment:
Do we need an interface? Why can't we just have a `Parameter` class? What's the value of the interface?

Reply:
`IParameter` is not necessary since there's only one impl for that interface. It was once useful when `Parameter` depends on `Json.Net`, which helps isolation implementation from the definition and makes the swapping of Json library eaiser. For now since `Parameter` has already switched to System.Text.Json, there's no immediate value for `IParameter` to exist.

I would still suggest to keep it though because all other classes (`Tuner` | `SearchSpace` | `Option`) depends on `IParameter`. And such isolation also helps make the improvement of `Parameter` easier. Some implementations in `Parameter` are quite hacky, and it's quite likely some day in the future we need to improve those implementations.

Reply:
I'm not convinced an interface makes isolation easier. Does this interface need to be public?

My concern is that once you ship a public interface, you can't change it. Having only one class implement an interface is kind of an anti-pattern.

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/interface

> ✔️ DO define an interface if you need some common API to be supported by a set of types that includes value types.

Here we only have 1 type.

Reply:
### Resolved